### PR TITLE
sync: Avoid crash on invalid state events

### LINKED ISSUE #2949: Not robust against invalid state event
I have a room with this `m.room.join_rules` state event:
```
{
  "content": {
    "join_rule": "publictest"
  },
  "origin_server_ts": 1613508609851,
  "sender": "@benoitx:matrix.org",
  "state_key": "",
  "type": "m.room.join_rules",
  "unsigned": {
    "replaces_state": "$FZOkUjJvHm8mTuUkJBDhTOS86xfq17rZuxo0MRKBwH8",
    "age": 83365036681
  }
```
which is not valid regarding the [Matrix specs](https://spec.matrix.org/v1.9/client-server-api/#mroomjoin_rules), but the server did not perform any validation when I sent it, so it's there and the SDK must accept it and ignore any invalid value.

When I send a message to this room, I have got this error in loop:
> matrix_sdk_ui::sync_service: Error while processing room list in sync service: SlidingSync failed: the enum variant JoinRule::_Custom cannot be serialized | crates/matrix-sdk-ui/src/sync_service.rs:283.

Then the app becomes unusable, because the sync is restarted automatically in case of error. This is another issue that we will have to handle on the app side.

Rageshake: https://github.com/element-hq/element-x-android-rageshakes/issues/1210

This is how we mitigated this on the Kotlin SDK: https://github.com/element-hq/element-android/pull/3062

### PR DESCRIPTION
**Pull request by timokoesters**
The problem was in matrix_sdk_sqlite::state_store::SqliteStateStore save_changes, which saved the RoomInfo to the disk by serializing it.

- [ ] Public API changes documented in changelogs (optional)

Fixes #2949

### CODE REVIEW

**Comment by timokoesters** [comment]
The bad events will now be saved as None, which falls back to the default for the event type. This may or may not be the correct behavior...

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Thanks for writing a fix. The approach isn't great, in the sense that the lib user might have been able to do something with the custom event, and we're losing it here.

Instead, could we try to serialize the `Raw<Event>`s in the `BaseRoomInfo`, have getters that automatically deserialize them, please? It shouldn't require a migration, in theory (because a serialized `Raw<T>` is the same as the serialize `T`, when the serialization mechanism is JSON). If it does require a migration, that would result in a logout in the EX apps, so it would be an easy way to try it out and make sure no migration is needed.

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 1]
```diff
@@ -1778,6 +1823,65 @@ mod tests {
         );
     }
 
+    #[async_test]
+    async fn test_save_bad_joinrules() {
```
Comment (timokoesters) [inline crates/matrix-sdk-base/src/rooms/normal.rs:1827]: This test is currently not testing what it's supposed to, because it runs in-memory instead of on sqlite or indexeddb. How can I change that?

File: crates/matrix-sdk-indexeddb/src/state_store/mod.rs [hunk 2]
```diff
@@ -741,37 +815,87 @@ impl_state_store!({
         if !changes.redactions.is_empty() {
             let state = tx.object_store(keys::ROOM_STATE)?;
             let room_info = tx.object_store(keys::ROOM_INFOS)?;
+            let raw_room_infos = tx.object_store(keys::RAW_ROOM_INFOS)?;
 
             for (room_id, redactions) in &changes.redactions {
+                // TODO: Load room version lazily, only if it is necessary
```
Comment (timokoesters) [inline crates/matrix-sdk-indexeddb/src/state_store/mod.rs:821]: I removed the lazy loading to make it simpler while implementing, but I should add it back in. The approach from the sqlite make_room_version closure doesn't work here because it's async

**Comment by timokoesters** [comment]
This PR is ready for another review. I've changed the fundamental approach, so now the raw events are stored in the database. I did a quick test using element x android (sqlite), but I could not test indexeddb yet and I haven't tested redactions yet.

File: crates/matrix-sdk-base/src/rooms/mod.rs [hunk 3]
```diff
@@ -116,6 +128,34 @@ pub struct BaseRoomInfo {
     pub(crate) notable_tags: RoomNotableTags,
 }
 
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub struct RawRoomInfo {
```
[the remark this answers is not in the document]
Comment (bnjbvr) [reply crates/matrix-sdk-base/src/rooms/mod.rs:132]: also, question: does it need to be public so that downstream implementations of the state store trait can reuse it? (If so, please indicate so in the comment.)

I wonder if it still needs that, with my suggestion above to use `serde(from / into)`.

File: crates/matrix-sdk-sqlite/src/state_store.rs [hunk 4]
```diff
@@ -1133,8 +1238,74 @@ impl StateStore for SqliteStateStore {
                             })
                     };
 
-                    let encoded_room_id = this.encode_key(keys::STATE_EVENT, &room_id);
-                    let mut room_version = None;
+                    // Clean up raw events we saved for roominfo
+                    if room_infos.get(room_id).is_some() {
```
Comment (bnjbvr) [inline crates/matrix-sdk-sqlite/src/state_store.rs:1242]: Why do we need this at all? :fearful: Do we need something similar for edits? If we actually have to keep this, can we commonize the code with the indexeddb implementation?

File: crates/matrix-sdk-indexeddb/src/state_store/mod.rs [hunk 5]
```diff
@@ -101,6 +102,7 @@ mod keys {
 
     pub const ROOM_STATE: &str = "room_state";
     pub const ROOM_INFOS: &str = "room_infos";
+    pub const RAW_ROOM_INFOS: &str = "raw_room_infos";
```
Comment (bnjbvr) [inline crates/matrix-sdk-indexeddb/src/state_store/mod.rs:105]: I don't think you'd need a new table, in both cases; the justification being that serializing `Raw<T>` with serde_json is the same as serializing `T` with serde_json. So you should be able to reuse the same table.

**Review by bnjbvr — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk-indexeddb/src/state_store/mod.rs (hunk 2 above)
Comment (timokoesters) [reply crates/matrix-sdk-indexeddb/src/state_store/mod.rs:760]: With the last set of changes, I had to make the sqlite version non-async too, to allow passing the room version into the base_info.handle_redaction function.

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 6]
```diff
@@ -850,8 +851,51 @@ pub struct RoomInfo {
     pub(crate) base_info: Box<BaseRoomInfo>,
 }
 
+/// This is the same as RoomInfo, except it contains RawBaseRoomInfo and implements Serialize.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RawRoomInfo {
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:856]: The sheer amount of code duplication scares me, here; is there a way to have an intermediate struct `CommonRoomInfo<T>`, which would implement `Serialize`, and then `RawRoomInfo` would be `CommonRoomInfo<RawBaseRoomInfo>` while `RoomInfo` would be `CommonRoomInfo<BaseRoomInfo>`?

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 7]
```diff
@@ -1127,21 +1188,24 @@ impl RoomInfo {
...
+            None => &GuestAccess::Forbidden,
         }
     }
 
     fn history_visibility(&self) -> &HistoryVisibility {
         match &self.base_info.history_visibility {
             Some(MinimalStateEvent::Original(ev)) => &ev.content.history_visibility,
-            _ => &HistoryVisibility::WorldReadable,
+            Some(MinimalStateEvent::Redacted(ev)) => &ev.content.history_visibility,
+            None => &HistoryVisibility::WorldReadable,
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:1200]: What's the motivation behind this change?

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 8]
```diff
@@ -1127,21 +1188,24 @@ impl RoomInfo {
...
         }
     }
 
     fn join_rule(&self) -> &JoinRule {
         match &self.base_info.join_rules {
             Some(MinimalStateEvent::Original(ev)) => &ev.content.join_rule,
-            _ => &JoinRule::Public,
+            Some(MinimalStateEvent::Redacted(ev)) => &ev.content.join_rule,
+            None => &JoinRule::Public,
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:1208]: Ditto?

File: crates/matrix-sdk-base/src/rooms/normal.rs (hunk 1 above)
Comment (bnjbvr) [reply crates/matrix-sdk-base/src/rooms/normal.rs:1853]: Did you address this comment? ^

File: crates/matrix-sdk-indexeddb/src/state_store/mod.rs [hunk 9]
```diff
@@ -623,10 +624,23 @@ impl_state_store!({
...
+
+                let mut raw_room_info = room_infos
+                    .get(&self.encode_key(keys::ROOM_INFOS, room_id))?
+                    .await?
+                    .and_then(|v| self.deserialize_event::<RawRoomInfo>(&v).ok())
+                    .unwrap_or_else(|| room_info.clone().into_raw_lossy());
+
+                if let Some(room_state_changes) = changes.state.get(room_id) {
+                    raw_room_info.base_info.extend_with_changes(&room_state_changes);
+                }
```
Comment (bnjbvr) [inline crates/matrix-sdk-indexeddb/src/state_store/mod.rs:639]: It seems it would be much simpler to throw away the previous `RawRoomInfo` and convert the `RoomInfo` struct into a `RawRoomInfo` struct, instead of applying a diff of changes to the previously-serialized `RawRoomInfo`? That way maybe we wouldn't even need `extend_with_changes` at all.

File: crates/matrix-sdk-indexeddb/src/state_store/mod.rs [hunk 10]
```diff
@@ -740,38 +754,55 @@ impl_state_store!({
...
+                    .and_then(|f| self.deserialize_event::<RoomInfo>(&f).ok())
+                    .and_then(|info| info.room_version().cloned())
+                    .unwrap_or_else(|| {
+                        warn!(?room_id, "Unable to find the room version, assume version 9");
+                        RoomVersionId::V9
+                    });
+
+                // Clean up raw events we saved for roominfo
+                if changes.room_infos.get(room_id).is_some() {
+                    // This roominfo has changed, so let's update the raw events
```
Comment (bnjbvr) [inline crates/matrix-sdk-indexeddb/src/state_store/mod.rs:773]: I don't understand why we need this :/ It seems that a redaction would cause a change to the `RoomInfo`, so we'd convert it into a `RawRoomInfo` before persisting it, and we wouldn't need any special handling of redacted events?

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Thanks for the new version of the PR! Here's another round of comments. I'm a bit scared by the code duplication, so let's see if we can lower it a bit.

File: crates/matrix-sdk-indexeddb/src/state_store/mod.rs (hunk 10 above)
Comment (bnjbvr) [reply crates/matrix-sdk-indexeddb/src/state_store/mod.rs:773]: Sorry, I think I understand it now. So as I've suggested in a private channel, the duplication of the code applying the redaction to events both in `BaseRoomInfo` and `RawRoomInfo` is a bit problematic, and a hindrance in the long term. I wonder if we could regenerate a `BaseRoomInfo` from the `RawRoomInfo` every time we received an update to the `RawRoomInfo`, instead? That would keep the redaction code in a single place and avoid duplication of concerns.

**Comment by bnjbvr** [comment]
This wasn't the path we wanted to take with this, so closing, we can revisit the approach later, as we still have an issue open describing the issue.
